### PR TITLE
WIP:  document Profile() class under _lab

### PR DESCRIPTION
**Pull request by kingrpaul**
@SimonBiggs the rst document for Profile() under _labs seems to be working pretty well. I added a reference in index.rst to get to it. It still needs polishing (WIP:) , but do you see any gross problems with the way this fits together?

Also, when this is complete, I would like to put a small GUI in front of Profile(), and start using it. I know enough Tkinter to think I can make it work for that purpose -- well enough, but not great. If you can suggest a free tool, with a short learning curve, that I might like better, I hope you will. 

Also, github suggests that I delete the branch you just merged into master. Any reason I should not do that?

Thanks in advance, and thanks for the review and quick merge on the last branch.

### CODE REVIEW

**Comment by SimonBiggs** [comment]
@kingrpaul I am considering using the following library to create GUIs within `pymedphys`:

https://github.com/SimonBiggs/scriptedforms/blob/master/README.md

You might find that helpful. I designed it with Medical Physics GUI creation in mind.

**Comment by SimonBiggs** [comment]
Once the branch is in master you can feel free to delete it.

**Comment by SimonBiggs** [comment]
One serious deal breaker is that the docs are no longer passing the build tests on the CI:

https://dev.azure.com/pymedphys/pymedphys/_build/results?buildId=564

**Comment by kingrpaul** [comment]
> 
> 
> @kingrpaul I am considering using the following library to create GUIs within `pymedphys`:
> 
> https://github.com/SimonBiggs/scriptedforms/blob/master/README.md
> 
> You might find that helpful. I designed it with Medical Physics GUI creation in mind.

Thank you @SimonBiggs. I will scope it out.

**Comment by kingrpaul** [comment]
> 
> 
> One serious deal breaker is that the docs are no longer passing the build tests on the CI:
> 
> https://dev.azure.com/pymedphys/pymedphys/_build/results?buildId=564

Oops. Didn't wait for the green check before asking. Understood that it would need to pass before merge.

**Comment by kingrpaul** [comment]
> 
> 
> Once the branch is in master you can feel free to delete it.

Thank you @SimonBiggs. Will do.

**Comment by kingrpaul** [comment]
@SimonBiggs would you take a look at this: 
https://github.com/pymedphys/pymedphys/runs/80375040

Fails the pipeline, complaining about one of the methods, but I'm having difficulty making sense of it. 
I would certainly appreciate any advice you can offer. 

- Paul

**Comment by SimonBiggs** [comment]
Somewhere it is looking for the following path and failing:

_labs/paulking/pymedphys._labs.paulking.profile.Profile.align_to

**Comment by kingrpaul** [comment]
@SimonBiggs I've gone through the pipeline failure best I can and I cannot find an explanation for the failure. The page looks good to me, as rendered. The error messages come and go from the desktop, but appear on github either way. 

I'm coming to conclude that the problem is with the pipeline. I don't think it's constructive to keep chasing this.  Going to move forward with the GUI.

Could you take a look at this and see if the documentation looks reasonable? 
https://github.com/pymedphys/pymedphys/commit/a08e284089b7c1728b0b767f055948812532d8a1

Much appreciated.

**Comment by SimonBiggs** [comment]
One addition I would highly appreciate would be in the docs on the website can the labs be nested under a heading of your name, not just as a directory, but also as a title. 

I think we also need sufficient warnings associated with the lab packages. So when users click on labs they are greeted by a description along the lines of:

"here are some experimental modules that have been written by both members of the community and maintainers of pymedphys. They have not yet undergone either sufficient testing, review, or refinement to be included within pymedphys proper. Consider everything under this module to be always at an "alpha" level of development. Consider that from any version to any version of pymedphys release these modules may change in such a way that breaks code that depends on it. Changes within the labs are not tracked within the changelogs."

"The labs are organised by primary author of each experiment, please click on an author to see the documentation for their current experiments."

Then after there have a list of each of our names that have something in the labs docs (just yours for now). Clicking on the name takes one to the respective lab docs.

**Comment by SimonBiggs** [comment]
Readthedocs, trying to build your branch, gives me the following:

```
writing... WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.align_to'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.cross_calibrate'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.from_lists'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.from_narrow_png'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.from_pulse'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.from_snc_profiler'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.from_tuples'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.get_edges'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.get_flatness'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.get_increment'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.get_symmetry'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.get_x'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.get_y'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.make_centered'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.make_flipped'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.make_normal_x'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.make_normal_y'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.make_symmetric'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.plot'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.resample_x'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.resample_y'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.slice_penumbra'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.slice_segment'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.slice_shoulders'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.slice_tails'
WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.slice_umbra'
```

**Comment by SimonBiggs** [comment]
So, both our live documentation site agrees with the pipeline

**Comment by SimonBiggs** [comment]
@kingrpaul when you're editing the docs do you use the following command:

```bash
sphinx-autobuild docs docs/_build/html
```

**Comment by kingrpaul** [comment]
> One addition I would highly appreciate would be in the docs on the website can the labs be nested under a heading of your name, not just as a directory, but also as a title.

OK, I think I know how to do that.

**Comment by kingrpaul** [comment]
> Readthedocs, trying to build your branch, gives me the following:
> ```
> writing... WARNING: toctree contains reference to nonexisting document '_labs/paulking/pymedphys._labs.paulking.profile.Profile.align_to'

Yes, that is more or let what I see as well. But, it is not the case. toctree points only to profile.rst and I have stripped profile.rst so that it does nothing but a single autoclass. The only indication as to the problem that I had was that I was trying to include an image. Pulling it out didn't help. There's nothing if-fy that I see in the way the webpage renders.

**Comment by SimonBiggs** [comment]
I haven't tested it, but notice how you're doing an auto module. Meaning
you want everything in that module printed. Notice how on other pages we've
said do automodule but don't include any methods. And then we have
selectively chosen which methods we want... Maybe that's the issue?

**Comment by SimonBiggs** [comment]
Given that warning are our indicator that something is wrong, we can't
leave warning in there even if the page looks right. Because the moment
warnings are left in there we no longer get informed about new warnings.

**Comment by kingrpaul** [comment]
Haven't given up, just at a loss for how to troubleshoot.  I did try using scriptedforms to create a user interface in front of the profile class, but had some challenges and am using tkinter, at present, which I am having better luck with than expected.

**Comment by SimonBiggs** [comment]
Awesome, go with tkinter. Go with what works.

Another tool I had forgotten about, but just remembered was:

https://github.com/chriskiehl/Gooey

**Comment by kingrpaul** [comment]
@SimonBiggs   
When this pipeline fails, it says it is looking for something in this location, which is not a location that I am supplying anywhere and is a location that does not exist. 
'_labs/paulking/pymedphys._labs.paulking.profile.Profile.align_to'
The elements of the path are duplicated. Is there anything in the way that sphinx was configured that would lead it to look this destination. I know that it is finding the path I supply because the webpage renders and if I change it, then the webpage does not render. 

Removing the methods from the documentation makes the desktop errors go away, but they still appear on github.

**Comment by kingrpaul** [comment]
> 
> 
> Awesome, go with tkinter. Go with what works.
> 
> Another tool I had forgotten about, but just remembered was:
> 
> https://github.com/chriskiehl/Gooey

Cool. Actually that looks really useful, though maybe not beefy enough for what I am doing. I found a description of how to embed a matplotlib plot into a tkinter window, and it's working pretty well. 
I've pushed this up to "the _lab" as a work in progress. Is it ok for GUI stuff to stay here? I can pull this out of pymedphys, if needed.

**Comment by SimonBiggs** [comment]
A google search for "autoclass nonexisting document" turned up the following:

https://stackoverflow.com/a/43237890/3912576

I think we need to change from numpydoc to napoleon.

File: docs/conf.py [hunk 1]
```diff
@@ -58,7 +58,7 @@ def setup(app):
     'sphinx.ext.autosummary',
     'sphinx.ext.graphviz',
     'matplotlib.sphinxext.plot_directive',
-    'numpydoc',
+    'sphinxcontrib.napoleon',
```
Comment (SimonBiggs) [inline docs/conf.py]: @Centrus007 due to [this bug](https://stackoverflow.com/a/43237890/3912576) I have changed from `numpydoc` to `sphinxcontrib.napoleon`.

File: src/pymedphys/dicom/_level2/dose.py [hunk 2]
```diff
@@ -358,11 +358,11 @@ def extract_dicom_patient_xyz(ds):
 
     References
     ----------
-    .. [1] "C.7.6.2.1.1 Image Position and Image Orientation",
+    .. [5] "C.7.6.2.1.1 Image Position and Image Orientation",
        "DICOM PS3.3 2016a - Information Object Definitions",
        http://dicom.nema.org/MEDICAL/dicom/2016a/output/chtml/part03/sect_C.7.6.2.html#sect_C.7.6.2.1.1
 
-    .. [2] O. McNoleg, "Generalized coordinate transformations for Monte Carlo
+    .. [6] O. McNoleg, "Generalized coordinate transformations for Monte Carlo
```
Comment (SimonBiggs) [inline src/pymedphys/dicom/_level2/dose.py]: @Centrus007 By swapping to `sphinxcontrib.napoleon` it complained of duplicate references in `dose.py`. So I changed that. Let me know if you'd like it a little different.

Cheers,
Simon

**Comment by SimonBiggs** [comment]
> Is it ok for GUI stuff to stay here? I can pull this out of pymedphys, if needed.

If it is just a part of your lab files it is just fine. We certainly need to make a few more decisions before GUI code finds its way into `pymedphys` proper.

**Comment by SimonBiggs** [comment]
@kingrpaul that's passing CI now :)

https://dev.azure.com/pymedphys/pymedphys/_build/results?buildId=690

**Comment by kingrpaul** [comment]
> @kingrpaul that's passing CI now :)
> 
> https://dev.azure.com/pymedphys/pymedphys/_build/results?buildId=690

Yes, it is. Much appreciated and much better. After merging this back into my local copy, I am now seeing some failures in areas that I don't edit. Assuming this is fallout from the documentation system change, not for me to try to fix.

**Comment by kingrpaul** [comment]
> > Is it ok for GUI stuff to stay here? I can pull this out of pymedphys, if needed.
> 
> If it is just a part of your lab files it is just fine. We certainly need to make a few more decisions before GUI code finds its way into `pymedphys` proper.

Thinking about this, I am leaning toward making a duplicate of Profile() outside pymedphys and keeping the GUI there while leaving the math and documentation inside. I would like to get Profile() inside pymedphys-proper. Seems, right now, like splitting it this way may be the best way to go. For now.

**Comment by SimonBiggs** [comment]
> I am leaning toward making a duplicate of Profile() outside pymedphys and keeping the GUI there while leaving the math and documentation inside.

Another method which might be better would be to create a clear API wrapper functions inside your labs, have your GUI only call the wrapper functions, and then as Profile is moved into pymedphys proper the _labs version with its wrappers can iteratively call the new version but still provide the same API to your GUI...

**Comment by SimonBiggs** [comment]
@kingrpaul before merging, I'll need to ask @Centrus007 to review the changes to his parts of the docs.

**Comment by kingrpaul** [comment]
Thank you.

**Comment by SimonBiggs** [comment]
You're welcome :), I reckon you should submit another pull request and put a little bit of a bio advertising yourself and your centre at the following page:

https://pymedphys.com/en/latest/labs/paulking/paulking.html

**Pull request merged**